Typescript error on theme import "Cannot find module..."

Because player.style doesn't currently export CommonJS modules, when importing player.style themes into a project using TypeScript, you might get an error that looks like this:

```
Cannot find module 'player.style/microvideo' or its corresponding type declarations.
```

The "fix" for this currently is to make sure that:

- You are importing the theme into a Javascript Module file and not a CommonJS file. Either by using the the package.json `"type": "module"` for the whole project. Or, the `.mts` extension for the file
- `module` in tsconfig is set to `nodenext`

--- a/package.json
+++ b/package.json
@@ -18,9 +18,34 @@
   ],
   "type": "module",
   "exports": {
-    "./*/react": "./themes/*/dist/react.js",
-    "./*.js": "./themes/*/dist/media-theme.js",
-    "./*": "./themes/*/dist/media-theme.js"
+    "./*/react": {
+      "types": "./themes/*/dist/react.d.ts",
+      "import": "./themes/*/dist/react.js",
+      "require": "./themes/*/dist/cjs/react.js",
+      "default": "./themes/*/dist/react.js"
+    },
+    "./*.js": {
+      "types": "./themes/*/dist/media-theme.d.ts",
+      "import": "./themes/*/dist/media-theme.js",
+      "require": "./themes/*/dist/cjs/media-theme.js",
+      "default": "./themes/*/dist/media-theme.js"
+    },
+    "./*": {
+      "types": "./themes/*/dist/media-theme.d.ts",
+      "import": "./themes/*/dist/media-theme.js",
+      "require": "./themes/*/dist/cjs/media-theme.js",
+      "default": "./themes/*/dist/media-theme.js"
+    }
+  },
+  "typesVersions": {
+    "*": {
+      "*/react": [
+        "./themes/*/dist/react.d.ts"
+      ],
+      "*": [
+        "./themes/*/dist/media-theme.d.ts"
+      ]
+    }
   },
   "workspaces": [
     ".",

--- a/scripts/build-theme/build.js
+++ b/scripts/build-theme/build.js
@@ -5,6 +5,7 @@ import { join, dirname } from 'node:path';
 import { realpath, mkdir, readFile, writeFile } from 'node:fs/promises';
 import { fileURLToPath } from 'node:url';
 import ejs from 'ejs';
+import * as esbuild from 'esbuild';
 
 const nodePath = await realpath(process.argv[1]);
 const modulePath = await realpath(fileURLToPath(import.meta.url));
@@ -70,6 +71,18 @@ export async function build() {
     // Copy code file to dist folder and replace vars.
     const reactCode = await readFile(join(dirname(modulePath), '/templates/react.js'), 'utf8');
     await writeFile(`./dist/react.js`, populate(reactCode, themeName));
+
+    // Build CJS files
+    await mkdir('./dist/cjs', { recursive: true });
+    await writeFile(`./dist/cjs/package.json`, JSON.stringify({ type: 'commonjs' }));
+
+    await esbuild.build({
+      entryPoints: ['./dist/media-theme.js', './dist/react.js'],
+      bundle: true,
+      format: 'cjs',
+      external: ['react', 'media-chrome', './media-theme.js'],
+      outdir: './dist/cjs',
+    });
   }
 }
 

--- a/scripts/build-theme/package.json
+++ b/scripts/build-theme/package.json
@@ -8,6 +8,7 @@
     "build-theme": "./build.js"
   },
   "dependencies": {
-    "ejs": "^3.1.10"
+    "ejs": "^3.1.10",
+    "esbuild": "^0.24.0"
   }
 }

--- a/themes/demuxed-2022/package.json
+++ b/themes/demuxed-2022/package.json
@@ -17,10 +17,31 @@
     "dist"
   ],
   "type": "module",
-  "main": "dist/media-theme.js",
+  "types": "./dist/media-theme.d.ts",
+  "main": "./dist/media-theme.js",
   "exports": {
-    ".": "./dist/media-theme.js",
-    "./react": "./dist/react.js"
+    ".": {
+      "types": "./dist/media-theme.d.ts",
+      "import": "./dist/media-theme.js",
+      "require": "./dist/cjs/media-theme.js",
+      "default": "./dist/media-theme.js"
+    },
+    "./react": {
+      "types": "./dist/react.d.ts",
+      "import": "./dist/react.js",
+      "require": "./dist/cjs/react.js",
+      "default": "./dist/react.js"
+    }
+  },
+  "typesVersions": {
+    "*": {
+      "react": [
+        "./dist/react.d.ts"
+      ],
+      "*": [
+        "./dist/media-theme.d.ts"
+      ]
+    }
   },
   "scripts": {
     "clean": "rimraf dist",

--- a/themes/instaplay/package.json
+++ b/themes/instaplay/package.json
@@ -17,10 +17,31 @@
     "dist"
   ],
   "type": "module",
-  "main": "dist/media-theme.js",
+  "types": "./dist/media-theme.d.ts",
+  "main": "./dist/media-theme.js",
   "exports": {
-    ".": "./dist/media-theme.js",
-    "./react": "./dist/react.js"
+    ".": {
+      "types": "./dist/media-theme.d.ts",
+      "import": "./dist/media-theme.js",
+      "require": "./dist/cjs/media-theme.js",
+      "default": "./dist/media-theme.js"
+    },
+    "./react": {
+      "types": "./dist/react.d.ts",
+      "import": "./dist/react.js",
+      "require": "./dist/cjs/react.js",
+      "default": "./dist/react.js"
+    }
+  },
+  "typesVersions": {
+    "*": {
+      "react": [
+        "./dist/react.d.ts"
+      ],
+      "*": [
+        "./dist/media-theme.d.ts"
+      ]
+    }
   },
   "scripts": {
     "clean": "rimraf dist",

--- a/themes/microvideo/package.json
+++ b/themes/microvideo/package.json
@@ -17,10 +17,31 @@
     "dist"
   ],
   "type": "module",
-  "main": "dist/media-theme.js",
+  "types": "./dist/media-theme.d.ts",
+  "main": "./dist/media-theme.js",
   "exports": {
-    ".": "./dist/media-theme.js",
-    "./react": "./dist/react.js"
+    ".": {
+      "types": "./dist/media-theme.d.ts",
+      "import": "./dist/media-theme.js",
+      "require": "./dist/cjs/media-theme.js",
+      "default": "./dist/media-theme.js"
+    },
+    "./react": {
+      "types": "./dist/react.d.ts",
+      "import": "./dist/react.js",
+      "require": "./dist/cjs/react.js",
+      "default": "./dist/react.js"
+    }
+  },
+  "typesVersions": {
+    "*": {
+      "react": [
+        "./dist/react.d.ts"
+      ],
+      "*": [
+        "./dist/media-theme.d.ts"
+      ]
+    }
   },
   "scripts": {
     "clean": "rimraf dist",

--- a/themes/minimal/package.json
+++ b/themes/minimal/package.json
@@ -17,10 +17,31 @@
     "dist"
   ],
   "type": "module",
-  "main": "dist/media-theme.js",
+  "types": "./dist/media-theme.d.ts",
+  "main": "./dist/media-theme.js",
   "exports": {
-    ".": "./dist/media-theme.js",
-    "./react": "./dist/react.js"
+    ".": {
+      "types": "./dist/media-theme.d.ts",
+      "import": "./dist/media-theme.js",
+      "require": "./dist/cjs/media-theme.js",
+      "default": "./dist/media-theme.js"
+    },
+    "./react": {
+      "types": "./dist/react.d.ts",
+      "import": "./dist/react.js",
+      "require": "./dist/cjs/react.js",
+      "default": "./dist/react.js"
+    }
+  },
+  "typesVersions": {
+    "*": {
+      "react": [
+        "./dist/react.d.ts"
+      ],
+      "*": [
+        "./dist/media-theme.d.ts"
+      ]
+    }
   },
   "scripts": {
     "clean": "rimraf dist",

--- a/themes/notflix/package.json
+++ b/themes/notflix/package.json
@@ -17,10 +17,31 @@
     "dist"
   ],
   "type": "module",
-  "main": "dist/media-theme.js",
+  "types": "./dist/media-theme.d.ts",
+  "main": "./dist/media-theme.js",
   "exports": {
-    ".": "./dist/media-theme.js",
-    "./react": "./dist/react.js"
+    ".": {
+      "types": "./dist/media-theme.d.ts",
+      "import": "./dist/media-theme.js",
+      "require": "./dist/cjs/media-theme.js",
+      "default": "./dist/media-theme.js"
+    },
+    "./react": {
+      "types": "./dist/react.d.ts",
+      "import": "./dist/react.js",
+      "require": "./dist/cjs/react.js",
+      "default": "./dist/react.js"
+    }
+  },
+  "typesVersions": {
+    "*": {
+      "react": [
+        "./dist/react.d.ts"
+      ],
+      "*": [
+        "./dist/media-theme.d.ts"
+      ]
+    }
   },
   "scripts": {
     "clean": "rimraf dist",

--- a/themes/reelplay/package.json
+++ b/themes/reelplay/package.json
@@ -17,10 +17,31 @@
     "dist"
   ],
   "type": "module",
-  "main": "dist/media-theme.js",
+  "types": "./dist/media-theme.d.ts",
+  "main": "./dist/media-theme.js",
   "exports": {
-    ".": "./dist/media-theme.js",
-    "./react": "./dist/react.js"
+    ".": {
+      "types": "./dist/media-theme.d.ts",
+      "import": "./dist/media-theme.js",
+      "require": "./dist/cjs/media-theme.js",
+      "default": "./dist/media-theme.js"
+    },
+    "./react": {
+      "types": "./dist/react.d.ts",
+      "import": "./dist/react.js",
+      "require": "./dist/cjs/react.js",
+      "default": "./dist/react.js"
+    }
+  },
+  "typesVersions": {
+    "*": {
+      "react": [
+        "./dist/react.d.ts"
+      ],
+      "*": [
+        "./dist/media-theme.d.ts"
+      ]
+    }
   },
   "scripts": {
     "clean": "rimraf dist",

--- a/themes/sutro-audio/package.json
+++ b/themes/sutro-audio/package.json
@@ -17,10 +17,31 @@
     "dist"
   ],
   "type": "module",
-  "main": "dist/media-theme.js",
+  "types": "./dist/media-theme.d.ts",
+  "main": "./dist/media-theme.js",
   "exports": {
-    ".": "./dist/media-theme.js",
-    "./react": "./dist/react.js"
+    ".": {
+      "types": "./dist/media-theme.d.ts",
+      "import": "./dist/media-theme.js",
+      "require": "./dist/cjs/media-theme.js",
+      "default": "./dist/media-theme.js"
+    },
+    "./react": {
+      "types": "./dist/react.d.ts",
+      "import": "./dist/react.js",
+      "require": "./dist/cjs/react.js",
+      "default": "./dist/react.js"
+    }
+  },
+  "typesVersions": {
+    "*": {
+      "react": [
+        "./dist/react.d.ts"
+      ],
+      "*": [
+        "./dist/media-theme.d.ts"
+      ]
+    }
   },
   "scripts": {
     "clean": "rimraf dist",

--- a/themes/sutro/package.json
+++ b/themes/sutro/package.json
@@ -17,10 +17,31 @@
     "dist"
   ],
   "type": "module",
-  "main": "dist/media-theme.js",
+  "types": "./dist/media-theme.d.ts",
+  "main": "./dist/media-theme.js",
   "exports": {
-    ".": "./dist/media-theme.js",
-    "./react": "./dist/react.js"
+    ".": {
+      "types": "./dist/media-theme.d.ts",
+      "import": "./dist/media-theme.js",
+      "require": "./dist/cjs/media-theme.js",
+      "default": "./dist/media-theme.js"
+    },
+    "./react": {
+      "types": "./dist/react.d.ts",
+      "import": "./dist/react.js",
+      "require": "./dist/cjs/react.js",
+      "default": "./dist/react.js"
+    }
+  },
+  "typesVersions": {
+    "*": {
+      "react": [
+        "./dist/react.d.ts"
+      ],
+      "*": [
+        "./dist/media-theme.d.ts"
+      ]
+    }
   },
   "scripts": {
     "clean": "rimraf dist",

--- a/themes/tailwind-audio/package.json
+++ b/themes/tailwind-audio/package.json
@@ -17,10 +17,31 @@
     "dist"
   ],
   "type": "module",
-  "main": "dist/media-theme.js",
+  "types": "./dist/media-theme.d.ts",
+  "main": "./dist/media-theme.js",
   "exports": {
-    ".": "./dist/media-theme.js",
-    "./react": "./dist/react.js"
+    ".": {
+      "types": "./dist/media-theme.d.ts",
+      "import": "./dist/media-theme.js",
+      "require": "./dist/cjs/media-theme.js",
+      "default": "./dist/media-theme.js"
+    },
+    "./react": {
+      "types": "./dist/react.d.ts",
+      "import": "./dist/react.js",
+      "require": "./dist/cjs/react.js",
+      "default": "./dist/react.js"
+    }
+  },
+  "typesVersions": {
+    "*": {
+      "react": [
+        "./dist/react.d.ts"
+      ],
+      "*": [
+        "./dist/media-theme.d.ts"
+      ]
+    }
   },
   "scripts": {
     "clean": "rimraf dist",

--- a/themes/vimeonova/package.json
+++ b/themes/vimeonova/package.json
@@ -17,10 +17,31 @@
     "dist"
   ],
   "type": "module",
-  "main": "dist/media-theme.js",
+  "types": "./dist/media-theme.d.ts",
+  "main": "./dist/media-theme.js",
   "exports": {
-    ".": "./dist/media-theme.js",
-    "./react": "./dist/react.js"
+    ".": {
+      "types": "./dist/media-theme.d.ts",
+      "import": "./dist/media-theme.js",
+      "require": "./dist/cjs/media-theme.js",
+      "default": "./dist/media-theme.js"
+    },
+    "./react": {
+      "types": "./dist/react.d.ts",
+      "import": "./dist/react.js",
+      "require": "./dist/cjs/react.js",
+      "default": "./dist/react.js"
+    }
+  },
+  "typesVersions": {
+    "*": {
+      "react": [
+        "./dist/react.d.ts"
+      ],
+      "*": [
+        "./dist/media-theme.d.ts"
+      ]
+    }
   },
   "scripts": {
     "clean": "rimraf dist",

--- a/themes/winamp/package.json
+++ b/themes/winamp/package.json
@@ -17,10 +17,31 @@
     "dist"
   ],
   "type": "module",
-  "main": "dist/media-theme.js",
+  "types": "./dist/media-theme.d.ts",
+  "main": "./dist/media-theme.js",
   "exports": {
-    ".": "./dist/media-theme.js",
-    "./react": "./dist/react.js"
+    ".": {
+      "types": "./dist/media-theme.d.ts",
+      "import": "./dist/media-theme.js",
+      "require": "./dist/cjs/media-theme.js",
+      "default": "./dist/media-theme.js"
+    },
+    "./react": {
+      "types": "./dist/react.d.ts",
+      "import": "./dist/react.js",
+      "require": "./dist/cjs/react.js",
+      "default": "./dist/react.js"
+    }
+  },
+  "typesVersions": {
+    "*": {
+      "react": [
+        "./dist/react.d.ts"
+      ],
+      "*": [
+        "./dist/media-theme.d.ts"
+      ]
+    }
   },
   "scripts": {
     "clean": "rimraf dist",

--- a/themes/yt/package.json
+++ b/themes/yt/package.json
@@ -17,10 +17,31 @@
     "dist"
   ],
   "type": "module",
-  "main": "dist/media-theme.js",
+  "types": "./dist/media-theme.d.ts",
+  "main": "./dist/media-theme.js",
   "exports": {
-    ".": "./dist/media-theme.js",
-    "./react": "./dist/react.js"
+    ".": {
+      "types": "./dist/media-theme.d.ts",
+      "import": "./dist/media-theme.js",
+      "require": "./dist/cjs/media-theme.js",
+      "default": "./dist/media-theme.js"
+    },
+    "./react": {
+      "types": "./dist/react.d.ts",
+      "import": "./dist/react.js",
+      "require": "./dist/cjs/react.js",
+      "default": "./dist/react.js"
+    }
+  },
+  "typesVersions": {
+    "*": {
+      "react": [
+        "./dist/react.d.ts"
+      ],
+      "*": [
+        "./dist/media-theme.d.ts"
+      ]
+    }
   },
   "scripts": {
     "clean": "rimraf dist",

